### DISCUSSION TITLE
line_style_doc program under X11: join type wrong

### DISCUSSION THREAD

**Discussion by wcout** [fltk/fltk#1017] [General]
Regarding #1016:

I notice, that under pure X11 (no wayland, no cairo) the join type is not respected:

![grafik](https://github.com/user-attachments/assets/4c571e10-1337-4cfc-99b7-b3db3253c64c)

I think it has to do with some special treatment of horizontal/vertical lines under X11 because of scaling.
Therefore I do not want to open an issue, though it is a regression compared to FLTK 1.3.

**Comment by Albrecht-S** [comment]
Thanks for reporting this. I noticed this as well, and it is similar - AFAICT - on Windows, tested both with and w/o `GDI+`. When I noticed this I left it open for a later research (I didn't want to open an issue yet).

I'm not sure that it's really a regression WRT 1.3. Using Greg's demo (1.3 lacks `Fl_Grid`) I get:
![image](https://github.com/user-attachments/assets/ba20fd04-5cb1-4f60-a686-447e77c11b85)

Notes:
- this screenshot from 1.3.10 (git) has been scaled to 200% by Gimp which does this, AFAICT, well, i.e. pixel-wise w/o anti-aliasing etc.
- I'm working on Debian 12 (Bookworm) under Gnome/Wayland, i.e. the screenshot was taken using branch-1.3 running under XWayland on Wayland. XWayland may affect the outcome in comparison to pure X
- admittedly the 1.4 screenshot by @wcout looks somewhat "worse" than my 1.3 screenshot, but both look wrong
- my tests with macOs looked much better (IMHO correct)
- I'm not entirely sure about Wayland and/or Cairo drawing on Linux, but ISTR that this was looking good as well

Leaving this report for others to improve the code if possible. @ManoloFLTK, maybe?

**Comment by erco77** [comment]
It's possible the FL_JOIN_ROUND isn't perceivable because I only used a 3 pixel thick line (IIRC) when generating the graphic.

If I [change the line thickness of test/line_style_docs.cxx](https://github.com/user-attachments/files/16336116/line_style_docs_v1.patch.txt) to be 8, I get better results for the caps, but would have to adjust the overall size of the demo too, as the dashed lines are now hard to perceive:

![image](https://github.com/user-attachments/assets/09d6a2b4-76ce-4278-9ed8-6c22754d2232)

That's with cairo+x11. But yeah, on pure X11, it's not drawing anything different for the "join" miter/round/bevel, confirming @wcout's result:
 
![image](https://github.com/user-attachments/assets/98519027-bed6-4637-a0c8-dabd43d9f6a8)

I don't know much about this subject really, just was updating the graphics : D
What /should/ these look like? Maybe I should check on a mac..

**Reply by erco77** [reply]
Ah, OK, here's the Mac results with the fat line modification (patch attached in the above post) which is perhaps the correct results we should use in the docs (but tweaking the app a bit so the dotted lines look right).

![image](https://github.com/user-attachments/assets/8d9618ea-ba89-48cc-91f6-bf3f1c3b91d4)

**Comment by Albrecht-S** [comment]
FWIW: there's also `test/line_style.cxx` to test the behavior interactively. No side-by-side graphics but no need to change source code.

**Comment by ManoloFLTK** [comment]
I've identified the cause and will commit a fix soon.

**Comment by ManoloFLTK** [comment]
Committed at 5e6c47f.

**Comment by wcout** [comment]
Why does it not work with the `line_style` demo?

![grafik](https://github.com/user-attachments/assets/61f18924-0af9-436b-a978-262811ff3377)

**Comment by erco77** [comment]
_Ha, oh wait, this is a "discussion" not and "issue".. they look the same, lol._
Re-opening because wcout is still seeing a problem.

**Reply by erco77** [reply]
Hmm, for the corners of the box to have round end caps, it seems you need to change Cap -> FL_CAP_ROUND, and not Join -> FL_JOIN_ROUND. That works for me in pure X11 mode:

![image](https://github.com/user-attachments/assets/359e8e23-6aff-4faf-82db-70bac91e2d3f)

_..I do see the filled circle overdraw a bit at 3 of the 4 corners; possibly because of how a thick line style affects circle drawing._

I guess how the flags affect drawing the rectangle depends on how, internally, an `fl_rect()` is drawn, as a continuous line with 4 vertices that have joins at each corner, or as 4 separate lines that each end in a cap. (The latter seems to be the case on X11)

**Comment by ManoloFLTK** [comment]
line_style is fixed with commit 3cc12d2.

**Comment by erco77** [comment]
Great! Thanks Manolo! Looks like you fixed everything; the overdraw of the circles, and the fl_rect() corners respond to the FL_JOIN_XXX flags now (instead of FL_CAP_XXX). Very nice!!
![image](https://github.com/user-attachments/assets/c369455e-0301-402c-805f-d0cbc1808256)
